### PR TITLE
Bigtable: add sync methods

### LINKED ISSUE #3635: BigtableDataClient is missing sync methods
Add below sync method in BigtableDataClient.

- [ ] readRow
- [ ] readRows
- [ ] sampleRowKeys
- [ ] mutateRow
- [ ] bulkMutateRows
- [ ] checkAndMutateRow
- [ ] readModifyWriteRow

### PR DESCRIPTION
**Pull request by elisheva-qlogic**
Fixes #3635 
Sync methods added in BigtableDataClient

### CODE REVIEW

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 1]
```diff
@@ -136,6 +137,54 @@ public static BigtableDataClient create(BigtableDataSettings settings) throws IO
...
+   * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create(instanceName)) {
+   *   String tableId = "[TABLE]";
+   *
+   *   Row row = bigtableDataClient.readRow(tableId, ByteString.copyFromUtf8("key"));
+   *   // Do something with row, for example, display all cells
+   *   System.out.println(row.getKey().toStringUtf8());
+   *   for(RowCell cell : row.getCells()) {
+   *     System.out.println("Family: " + cell.getFamily() + "   Qualifier: " + cell.getQualifier().toStringUtf8() + "   Value: " + cell.getValue().toStringUtf8());
+   *   }
+   * }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: Please update the docs for all sync methods with error handling info:
- In the sample code, add a catch clause mirroring the async variant.
- Add `@throws ApiException when a serverside error occurs `  to the end of the javadoc

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 2]
```diff
@@ -136,6 +137,54 @@ public static BigtableDataClient create(BigtableDataSettings settings) throws IO
...
+   * <p>Sample code:
+   *
+   * <pre>{code
+   * InstanceName instanceName = InstanceName.of("[PROJECT]", "[INSTANCE]");
+   * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create(instanceName)) {
+   *   String tableId = "[TABLE]";
+   *
+   *   Row row = bigtableDataClient.readRow(tableId, "key");
+   *   // Do something with row, for example, display all cells
+   *   System.out.println(row.getKey().toStringUtf8());
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: Thanks for adding details about using the row. Please add null check to show that the returned value can be null if the row isn't found. And consider making the async variant consistent

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 3]
```diff
@@ -547,6 +639,28 @@ public BulkMutationBatcher newBulkMutationBatcher() {
    * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create(instanceName)) {
    *   BulkMutation batch = BulkMutation.create("[TABLE]");
    *   for (String someValue : someCollection) {
+   *     batch.add("[ROW KEY]", Mutation.create().setCell("[FAMILY NAME]", "[QUALIFIER]", "[VALUE]"));
+   *   }
+   *   bigtableDataClient.bulkMutateRows(batch);
+   * }
+   * }</pre>
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: In addition to `@throws ApiException` also please add `@throws MutateRowsException if any of the entries failed to be applied.`

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java (hunk 2 above)
Comment (elisheva-qlogic) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: For making the async variant cosistent - I tried out the async example and onSuccess isn't called anyways if the row is null

Comment (igorbernstein2) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: Can you clarify what you mean? Here is an integration test that shows it does work:
https://github.com/googleapis/google-cloud-java/pull/3870

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 4]
```diff
@@ -136,6 +137,66 @@ public static BigtableDataClient create(BigtableDataSettings settings) throws IO
...
+   *   // Do something with row, for example, display all cells
+   *   if(row != null) {
+   *     System.out.println(row.getKey().toStringUtf8());
+   *     for(RowCell cell : row.getCells()) {
+   *       System.out.println("Family: " + cell.getFamily() + "   Qualifier: " + cell.getQualifier().toStringUtf8() + "   Value: " + cell.getValue().toStringUtf8());
+   *     }
+   *   }
+   * } catch(ApiException e) {
+   *    e.printStackTrace();
+   *    }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: please fix whitespace

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 5]
```diff
@@ -136,6 +137,66 @@ public static BigtableDataClient create(BigtableDataSettings settings) throws IO
...
+   *   // Do something with row, for example, display all cells
+   *   if(row != null) {
+   *     System.out.println(row.getKey().toStringUtf8());
+   *      for(RowCell cell : row.getCells()) {
+   *        System.out.println("Family: " + cell.getFamily() + "   Qualifier: " + cell.getQualifier().toStringUtf8() + "   Value: " + cell.getValue().toStringUtf8());
+   *      }
+   *   }
+   * } catch(ApiException e) {
+   *    e.printStackTrace();
+   *    }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: please fix whitespace

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 6]
```diff
@@ -374,6 +435,33 @@ public void readRowsAsync(Query query, ResponseObserver<Row> observer) {
...
+   * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create(instanceName)) {
+   *   String tableId = "[TABLE_ID]";
+   *
+   *   List<KeyOffset> keyOffsets = bigtableDataClient.sampleRowKeys(tableId);
+   *   for(KeyOffset keyOffset : keyOffsets) {
+   *   // Do something with keyOffset
+   *   }
+   * } catch(ApiException e) {
+   *    e.printStackTrace();
+   *   }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: please fix whitespace

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 7]
```diff
@@ -447,6 +535,30 @@ public void readRowsAsync(Query query, ResponseObserver<Row> observer) {
...
+   * <pre>{@code
+   * InstanceName instanceName = InstanceName.of("[PROJECT]", "[INSTANCE]");
+   * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create(instanceName)) {
+   *   RowMutation mutation = RowMutation.create("[TABLE]", "[ROW KEY]")
+   *     .setCell("[FAMILY NAME]", "[QUALIFIER]", "[VALUE]");
+   *
+   *   bigtableDataClient.mutateRow(mutation);
+   * } catch(ApiException e) {
+   *    e.printStackTrace();
+   *   }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: please fix whitespace

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 8]
```diff
@@ -547,6 +659,36 @@ public BulkMutationBatcher newBulkMutationBatcher() {
...
    *   for (String someValue : someCollection) {
+   *     batch.add("[ROW KEY]", Mutation.create().setCell("[FAMILY NAME]", "[QUALIFIER]", "[VALUE]"));
+   *   }
+   *   bigtableDataClient.bulkMutateRows(batch);
+   * } catch(ApiException e) {
+   *    e.printStackTrace();
+   *   }
+   *   catch(MutateRowsException e) {
+   *    e.printStackTrace();
+   *   }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: please fix whitespace

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 9]
```diff
@@ -594,6 +736,33 @@ public BulkMutationBatcher newBulkMutationBatcher() {
...
+   *     .condition(FILTERS.value().regex("old-value"))
+   *     .then(
+   *       Mutation.create()
+   *         .setCell("[FAMILY]", "[QUALIFIER]", "[VALUE]")
+   *       );
+   *
+   *   Boolean result = bigtableDataClient.checkAndMutateRow(mutation);
+   * } catch(ApiException e) {
+   *    e.printStackTrace();
+   *   }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: please fix whitespace

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java [hunk 10]
```diff
@@ -663,6 +832,33 @@ public BulkMutationBatcher newBulkMutationBatcher() {
...
+   * InstanceName instanceName = InstanceName.of("[PROJECT]", "[INSTANCE]");
+   * try (BigtableDataClient bigtableDataClient = BigtableDataClient.create(instanceName)) {
+   *   ReadModifyWriteRow mutation = ReadModifyWriteRow.create("[TABLE]", "[KEY]")
+   *     .increment("[FAMILY]", "[QUALIFIER]", 1)
+   *     .append("[FAMILY2]", "[QUALIFIER2]", "suffix");
+   *
+   *   Row success = bigtableDataClient.readModifyWriteRow(mutation);
+   * } catch(ApiException e) {
+   *    e.printStackTrace();
+   *   }
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: please fix whitespace

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java [hunk 11]
```diff
@@ -131,6 +136,48 @@ public void proxyReadRowStrAsyncTest() {
                 .build());
   }
 
+  @Test
+  public void readRowTest() {
+    Mockito.when(mockReadRowsCallable.first().futureCall(any(Query.class)))
+            .thenReturn(ApiFutures.immediateFuture(Row.create(ByteString.copyFromUtf8("fake-row-key"), Collections.emptyList())));
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java]: this line looks a bit long, please make sure to keep lines < 120 characters

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java [hunk 12]
```diff
@@ -131,6 +136,48 @@ public void proxyReadRowStrAsyncTest() {
...
+                            .setTableName("projects/fake-project/instances/fake-instance/tables/fake-table")
+                            .setAppProfileId("fake-profile")
+                            .setRows(RowSet.newBuilder().addRowKeys(ByteString.copyFromUtf8("fake-row-key")))
+                            .build());
+  }
+
+  @Test
+  public void readRowStrTest() {
+    Mockito.when(mockReadRowsCallable.first().futureCall(any(Query.class)))
+            .thenReturn(ApiFutures.immediateFuture(Row.create(ByteString.copyFromUtf8("fake-row-key"), Collections.emptyList())));
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java]: line is too long

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java [hunk 13]
```diff
@@ -160,6 +207,13 @@ public void proxySampleRowKeysTest() {
     Mockito.verify(mockSampleRowKeysCallable).futureCall("fake-table");
   }
 
+  @Test
+  public void sampleRowKeysTest() {
+    Mockito.when(mockSampleRowKeysCallable.futureCall(any(String.class))).thenReturn(ApiFutures.immediateFuture(Collections.emptyList()));
```
Comment (igorbernstein2) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java]: long line

File: google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java (hunk 2 above)
Comment (elisheva-qlogic) [reply google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/BigtableDataClient.java]: I added the null check for the async variant

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java [hunk 14]
```diff
@@ -168,13 +225,26 @@ public void proxyMutateRowCallableTest() {
...
+                    .setCell("some-family", "fake-qualifier", "fake-value");
 
     bigtableDataClient.mutateRowAsync(request);
     Mockito.verify(mockMutateRowCallable).futureCall(request);
   }
 
+  @Test
+  public void mutateRowTest() {
+    Mockito.when(mockMutateRowCallable.futureCall(any(RowMutation.class)))
+            .thenAnswer((Answer) (invocationOnMock) -> ApiFutures.immediateFuture(Empty.getDefaultInstance()));
```
Comment (chingor13) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java]: Unfortunately, we can't use lambda's yet as we are keeping Java 7 source compatibility.

File: google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java [hunk 15]
```diff
@@ -187,6 +257,21 @@ public void proxyBulkMutatesRowTest() {
     Mockito.verify(mockBulkMutateRowsCallable).futureCall(request);
   }
 
+  @Test
+  public void bulkMutatesRowTest() {
+    Mockito.when(mockBulkMutateRowsCallable.futureCall(any(BulkMutation.class)))
+            .thenAnswer((Answer) (invocationOnMock) -> ApiFutures.immediateFuture(Empty.getDefaultInstance()));
```
Comment (chingor13) [inline google-cloud-clients/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/BigtableDataClientTest.java]: Same thing on lambdas.

**Comment by igorbernstein2** [comment]
Build issues should be fixed now.

**Comment by chingor13** [comment]
Spanner and Logging IT test failures unrelated.

**Pull request merged**